### ISSUE TITLE
memory leaking

### ISSUE THREAD

**Issue by tophe** [gographics/imagick#327]
Hi Justin, 

I have an issue with a long running process, the process have consume almost 50Gb of Ram, but the go runtime says it was using 5Mb only.

So I have extract the code in a small demo, https://github.com/tophe/imgdemo, if you try it, you can see that the system memory grow ever and ever when you used the resize handler. (try to run "ruby  test.rb" and look via "ps afux| grep -v grep | grep imgd"), the goruntime don't see that memory usage.

I have remarked the CoalesceImages(), because it is very slow, it use even more memory and usualy it crash the go vm after some hits.

is it a go code probleme, I use Destroy() in the resize method, or is it a bug in the binding?

thank you for your help

**Comment by justinfx** [comment]
Hi. I'll try to have a look. 

Just FYI, this isn't the "Justin" project. I'm just someone who used to be a user of this library many many years ago, then stopped using it, but ended up being the only one who helps look after it 😉
I always hope others will jump in and fix things. Especially those who have a stake in keeping the bindings working.

**Comment by justinfx** [comment]
@tophe there is something strange in your repro. Your Dockerfile says it is building ImageMagick 7.1.1-36, but your go.mod file says it is using imagick.v2 which is for ImageMagick6. How are you getting that to work properly? Are you picking up the ImageMagick version you expect?

That being said, I looked over the repro and I think maybe it is still too complex. I've reduced it to a much simpler single executable, to get rid of the http server and external ruby script loop.

```go
package main

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"runtime"
	"sync"
	"time"

	"gopkg.in/gographics/imagick.v3/imagick"
)

func main() {
	fmt.Println("starting with cpus", runtime.NumCPU())
	imagick.Initialize()
	defer imagick.Terminate()

	var wg sync.WaitGroup
	for t := 0; t < runtime.NumCPU(); t++ {
		wg.Add(1)

		go func(id int) {
			defer wg.Done()
			for n := 0; n < 1000; n++ {
				fmt.Printf("g:%d, n:%d\n", id, n)
				if n%100 == 0 {
					fmt.Println("waiting 2 seconds")
					runtime.GC()
					time.Sleep(2 * time.Second)
				}
				resize()
			}
		}(t)
	}

	wg.Wait()
	fmt.Println("done")
}

func resize() {
	originalWand := imagick.NewMagickWand()
	finalWand := imagick.NewMagickWand()

	err := originalWand.ReadImage("./examples/tx.gif")
	if err != nil {
		log.Fatal(err)
	}

	numberImages := originalWand.GetNumberImages()

	for i := 0; i < int(numberImages); i++ {
		originalWand.SetIteratorIndex(i)
		mw := imagick.NewMagickWand()
		mw = originalWand.GetImage()

		//err = mw.ResizeImage(128, 105, imagick.FILTER_LANCZOS, 0)
		err = mw.ResizeImage(128, 105, imagick.FILTER_LANCZOS)
		if err != nil {
			panic(err)
		}

		if err = finalWand.AddImage(mw); err != nil {
			panic(err)
		}
		mw.Destroy()

	}

	originalWand.Destroy()

	var data []byte
	data, err = finalWand.GetImagesBlob()
	if err != nil {
		panic(err)
	}
	finalWand.Destroy()

	reader := bytes.NewReader(data)
	_, err = io.Copy(io.Discard, reader)
	if err != nil {
		panic(err)
	}
}
```

I tested this in the "golang:1.22.5" container, limiting the container to 4gb of memory, and it has 10 cpus available. The memory increased to a max at one point of 1.5GB, but I would observe it going up and down, as low as maybe 800MB. So it seemed to find its reasonable usage for 10 parallel resizes happening. 

Can you give this repro a test? I think it confirms that it is likely not a problem in the actual Imagick bindings, otherwise I should see it happening here. There are factors in your repro such as external delegate plugins for the images. 

This seems like a duplicate of an older issue #136. Can you give this issue a review as well? There are maybe things you can set in the "policy.xml" file.

**Comment by tophe** [comment]
hello Justin,

Thank you for your help.

I use the dockerfile for running the code with the binding v3 and I run in local with v2.

I tried your version, and I see a memory consumption off almost 1gb in imagick v2, the memory grow slowly, it seem ok.

I read the thread you mention, in production I use a bounded image workers limited to NumCpu/2 goroutines.
I use a  modified policy.xml, but it is only to raise the memory maximum to handle huge images.

I rework on the test file, and it seems to run with the same amount of memory as your code.
I thing, I was alerted by the VSZ fast growing but it seem that the RSS usage is’nt growing so fast.

In the production, I have an other action with is needed to resize, I have removed if from the small example for simplification purpose.

```
numberImages := originalWand.GetNumberImages()
if numberImages > 1 {
   // rebuild full frame in animation
   originalWand = originalWand.CoalesceImages()
}


```

when I added that line, and I relaunch your code snippet, I see a RSS memory consumption that grow rapidly to 2.5Gb after some seconds, and the the system crash.

Is it possible that the CoalesceImages() was the heart of the problem ?

Can you reproduce the problem, and is it a binding probleme ?

**Comment by justinfx** [comment]
@tophe thanks for confirming that your issue was looking at virtual memory and not RSS. Good to know that with my repro code, you observed a normal memory growth that is not a leak.

Regarding your issue with `CoalesceImages`, you can see how simple the source code is in the bindings: https://github.com/gographics/imagick/blob/v3.7.0/imagick/magick_wand_image.go#L275
All it does is make the call into C and wrap the results in a new wand instance. That being said, I was able to reproduce your quick memory growth by adding the following to the repro code:

```go
numberImages := originalWand.GetNumberImages()

if numberImages > 1 {
	originalWand = originalWand.CoalesceImages()
}
```

If I run this with a single thread, it quickly gets up to my container 4GB limit and crashes.

If you notice that `CoalesceImages` returns a NEW wand, we can see how this could be leaking original wand it is replacing. I made the following change to the code:

```go
numberImages := originalWand.GetNumberImages()

if numberImages > 1 {
	newWand := originalWand.CoalesceImages()
	originalWand.Destroy()
	originalWand = newWand
}
```

When I run this again with 1 thread, it made it through 1000 iterations and never got above 915MB. I noticed it would slowly go up, and back down a little, over time.

Can you confirm if this change fixes your issue? If so, we could close the ticket as resolved.

**Comment by tophe** [comment]
Oh, thank you justin !

I just run the test, and it run without any leak!

That's was just a bad usage of the CoalesceImages() func !
So it is the same thing with 
```
	if imgFormat == "GIF" {
		finalWand = finalWand.OptimizeImageLayers()
	}
```
I need to rewrite with

```
	if imgFormat == "GIF" {
                tmpWand := finalWand.OptimizeImageLayers()
                finalWand.Destroy()
                finalWand = tmpWand
 
	}
```

I take a closer look to my code, to fix other issue, with wand returning a wand.
that was a good code lesson thank you. 

I close the issue

**Comment by justinfx** [comment]
That's ok. It probably wasn't very obvious and it might only happen in a loop. Enjoy!

**Comment by tophe** [comment]
hi justin,

it seems that this correction dont work with the CoalesceImages()

I have made lots of test with a very large animated gif of 46Mb 360 frames.
and i have a memory leak as soon as I use the CoalesceImages().

I try to resume the simplest test you can do to reproduce the problem.

originalWand := imagick.NewMagickWand()
newWand := imagick.NewMagickWand()
#memory < 1Gb
err = originalWand.ReadImageBlob(buf.Bytes())
if err != nil {
	return
}
#memory 6Gb

originalWand.Destroy()
newWand.Destroy()
return

# memory < 1Gb

and then


originalWand := imagick.NewMagickWand()
newWand := imagick.NewMagickWand()
#memory < 1Gb
err = originalWand.ReadImageBlob(buf.Bytes())
if err != nil {
	return
}
#memory 6Gb
if numberImages > 1 {
	newWand = originalWand.CoalesceImages()
# memory 11Gb	
}

originalWand.Destroy()
# memory 5Gb
newWand.Destroy()
# memory 5Gb
return

# memory 5Gb

leaking 5Gb !

as you point to c code, it seem that the binding only call the c function, so I don't know if it is a go binding problem or an imagemagick problem.
have you any idea, to correct that problem ?

you can try, with that gif, a big gif example https://upload.wikimedia.org/wikipedia/commons/d/da/A_weather_balloon_exploding.gif
it leak 1Gb per call.

**Comment by justinfx** [comment]
It's only a leak if the RSS continues to increase as you repeat the same calls, and never reduces.
Your example shows it goes up to 11GB and then reduces and holds at 5GB. If you continue making those same calls, does it keep increasing and never reducing? The ImageMagick C library may be holding on to memory in a pool.
Since it is not the Go bindings holding the memory, you would need to use the ImageMagick debug logging environment variables (and maybe a C profiler) to see where the memory is being held.

**Comment by tophe** [comment]
hi justin,

I made a small bunch of code to reproduce the problem:
```
package main

import (
	"fmt"
	"gopkg.in/gographics/imagick.v3/imagick"
	"runtime"
	"time"
)

// get gif
// wget https://upload.wikimedia.org/wikipedia/commons/d/da/A_weather_balloon_exploding.gif

func main() {
	fmt.Println("starting with cpus", runtime.NumCPU())
	imagick.Initialize()
	defer imagick.Terminate()

	for i := 0; i < 10; i++ {
		coalesce(i)
	}

	time.Sleep(5 * time.Minute)
	fmt.Println("done")
}

func coalesce(run int) {
	fmt.Printf("starting coalesce : %v\n", run)
	originalWand := imagick.NewMagickWand()
	err := originalWand.ReadImage("./A_weather_balloon_exploding.gif")
	//err := originalWand.ReadImage("./c.gif")
	if err != nil {
		return
	}
	newWand := originalWand.CoalesceImages()
	fmt.Printf("GetLastError(): %v\n", newWand.GetLastError())
	originalWand.Destroy()
	newWand.Destroy()
	return

}
```

then I run it and I read the memory usage during the sleep.
sometime it end using  > 1Gb,  sometime it end freeing all it's memory ending with 19M.
each time I run it I have a new result sometime it use > 4.6Gb, sometime 2.2Gb.
that's almos strange.

and now, if I change 

```
	for i := 0; i < 10; i++ {
		go coalesce(i)
		time.Sleep(5 * time.Second)
	}
```
	
because it closer to my execution workflow, then it ends with 7Gb used sometime 4Gb, it change at each execution.	
that memory is wasted.

any idea ?

**Comment by justinfx** [comment]
Well yes if you do 10 in parallel you are going to greatly increase the memory consumed by ImageMagick. That part is known. But if you only use a single goroutine and you see the memory keep increasing then that is different than it going back down again. I'm less concerned with the max memory used and more concerned with whether it keeps rising and never frees when the work is done.

**Comment by justinfx** [comment]
Here is a modified version of your test:

```go
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"

	"gopkg.in/gographics/imagick.v3/imagick"
)

// get gif
// wget https://upload.wikimedia.org/wikipedia/commons/d/da/A_weather_balloon_exploding.gif

func main() {
	fmt.Println("starting with cpus", runtime.NumCPU())
	imagick.Initialize()
	defer imagick.Terminate()

	const (
		Parallel = 1 // max goroutines at one time
		Loops    = 3
		PerLoop  = 10
	)

	for range Loops {
		var wg sync.WaitGroup
		wg.Add(PerLoop)

		ch := make(chan struct{}, Parallel)
		for i := 0; i < PerLoop; i++ {
			i := i
			ch <- struct{}{}
			go func() {
				defer func() {
					<-ch
					wg.Done()
				}()
				coalesce(i)
			}()
		}

		wg.Wait()
		fmt.Println("sleeping")
		time.Sleep(20 * time.Second)
	}

	fmt.Println("done")
}

func coalesce(run int) {
	fmt.Printf("starting coalesce : %v\n", run)
	originalWand := imagick.NewMagickWand()
	err := originalWand.ReadImage("./examples/A_weather_balloon_exploding.gif")
	if err != nil {
		panic(err)
	}
	newWand := originalWand.CoalesceImages()
	if e := newWand.GetLastError(); e != nil {
		fmt.Printf("GetLastError(): %v\n", e)
	}
	originalWand.Destroy()
	newWand.Destroy()
	return

}
```

After each loop is done, when it is sleeping, I do see the memory decrease a bit. And then the next loop uses the same max memory as the previous loop. If you increase the `Parallel` count, you will see it use a lot more max memory and have a much higher resting memory usage.

**Comment by tophe** [comment]
hi Justin,

the problem in my use case is the resting memory witch is growing rapidly, and never release.
I made some test with the command line magick and I see i use the same amount of memory than the go api.
using magick a.gif -coalesce b.gif.

it seem to came from the image magick pixel cache. the memory exhaustion can me mitigated with the policy.xml, reducing the pixel cache memory.
I made some test with a 360 frames animated git with use more than 11Gb during coalesce. If the image doesn't fit in the pixel cache im use memory mapped file with tmp file.



> magick -list resource
> Resource limits:
>   Width: 32000P
>   Height: 32000P
>   Area: 1GP
>   List length: unlimited
>   Memory: 4GiB
>   Map: 0B
>   Disk: unlimited
>   File: 768
>   Thread: 31
>   Throttle: 0
>   Time: unlimited

with this configuration the system have a lesser resting memory, that is better for my use case.
at each call, the memory raise 2Gb, with is a lot lesser than before.

but the performance suffer from this configuration.
the better solution should be to limit the memory, for each image sequence > 1 images.
I have see it is possible with the binding in the documentation with MagickWand.SetResourceLimit().

but in the binding 7.x, i see than the function have moved at the package level. So it cannot be fixed for each magickWand.
and so I is not goroutine aware, and it unsure how it work in a concurrent environment.

lowering memory help me a lot.
thank you.
